lang/py/src/avro/avro_builder.py

```diff
@@ -0,0 +1,238 @@
+# -*- coding: utf-8 -*-
+import copy
+
+from avro import schema
+
+
+class AvroBuildInvalidOperation(Exception):
+    pass
+
+
+class AvroSchemaBuilder(object):
+    """
+    AvroSchemaBuilder creates json-formatted Avro schemas. It has `create_*`
+    function for each primitive type to create primitive types. To create a
+    complex type, start with corresponding `begin_*` function and finish it
+    with `end` function.
+
+    It defers the schema validation until the end of schema building. When
+    the schema building ends, it constructs the corresponding schema object
+    which will validate the the syntax of the Avro json object.
+
+    Usage: (unit tests cover more usages)
+      ab = AvroSchemaBuilder()
+
+      - building a record schema:
+        ab.begin_record('user', namespace='yelp')
+        ab.add_field('id', ab.create_int())
+        ab.add_field(
+            'fav_color',
+            ab.begin_enum('color_enum', ['red', 'blue']).end()
+        )
+        record = ab.end()
+
+      - building an enum schema:
+        enum_schema = ab.begin_enum('color_enum', ['red', 'blue']).end()
+
+      - building an array schema:
+        array_schema = ab.begin_array(ab.create_string()).end()
+    """
+
+    def __init__(self):
+        self._schema_json = None  # current avro schema in build
```

Comment:
I see your concern. I agree that it is probably more straight forward if all the functions are static functions that each returns an avro schema and let users to compose the nested schemas themselves. My idea is to allow users to build a schema without needing temp variables to hold the intermediate schemas if they'd like to. Also, it allows the validation to happen in the end of final schema instead of each intermediate schema. It's a good point though.